grassroots-backend/src/grassroots-shared/Permission.ts

```diff
@@ -0,0 +1,81 @@
+import {
+  AbilityBuilder,
+  createMongoAbility,
+  MongoAbility,
+  subject,
+} from "@casl/ability";
+
+import { CASLSubjects } from "@shared/CASLSubjects";
+
+export enum Permission {
+  VIEW_CONTACTS = "VIEW_CONTACTS",
+  MANAGE_CONTACTS = "MANAGE_CONTACTS",
+  MANAGE_USERS = "MANAGE_USERS",
+}
+
+export type CASLAction = "read" | "edit";
+
+export function buildCASLSubject<K extends keyof CASLSubjects>(
+  k: K,
+  x: Omit<CASLSubjects[K], "__caslSubjectType">,
+): CASLSubjects[K] {
+  // Type inference isn't quite clever enough here.
+  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+  return {
+    ...x,
+    __caslSubjectType: k,
+  } as unknown as CASLSubjects[K];
+}
+
+type CASLSubjectUnion = CASLSubjects[keyof CASLSubjects];
+
+// MongoAbility needs to know both about the string types (keyof CASLSubjects) and the object structure (CASLSubjects).
+// It maps from strings to object types via `detectSubjectType` below.
+// TODO: don't export once we have better tests.
+export type AppAbility = MongoAbility<
+  [CASLAction, keyof CASLSubjects | CASLSubjectUnion]
+>;
+
+// CASL uses "can" both to define abilities and query them.
+// Note that this is the method for querying.
+export function can(
+  ability: AppAbility,
+  action: CASLAction,
+  type: keyof CASLSubjects,
+  object: CASLSubjectUnion,
+): boolean {
+  return ability.can(action, subject(type, object));
+}
+
+export function permissionsToCaslAbilities(
```

Comment:
This should probably be refactored out to a separate file which only declares the rules.